woocommerce/woocommerce
# Add requesting state, actions, and selectors to CRUD data stores

### All Submissions:






### Changes proposed in this Pull Request:

This adds selectors to allow us to check if a dispatched action has completed or not.  We'll need this to check if products are still being generated in the variations component in order to show a spinner.

This PR can optionally be split into 2, moving the work for `hasFinishedRequest` and `isRequesting` to another PR, but I suspect most of the LOC are due to copy/pasting similar reducer actions.

Closes #36189  .






### How to test the changes in this Pull Request:



1.  Open any WooCommerce Admin Reactified page
2. In your browser's console, copy and paste this replacing `37` with a product category ID in your store and hit enter
```
const testActionRequestStatus = () => {
    wp.data.dispatch('experimental/wc/admin/products/categories').updateProductCategory(37, {name: 'New name'}).then( () => {
        console.log( 'Request complete' );
        console.log( wp.data.select('experimental/wc/admin/products/categories').hasFinishedRequest('updateProductCategory',[37, {name: 'New name'}]) );
        console.log( wp.data.select('experimental/wc/admin/products/categories').isRequesting('updateProductCategory',[37, {name: 'New name'}]) );
    } );
    console.log( wp.data.select('experimental/wc/admin/products/categories').hasFinishedRequest('updateProductCategory',[37, {name: 'New name'}]) );
    console.log( wp.data.select('experimental/wc/admin/products/categories').isRequesting('updateProductCategory',[37, {name: 'New name'}]) );
}
testActionRequestStatus();
```
3. You should see the following after the request has completed:
```
false
true
Request complete
true
false
```
4. You can also use full ID queries to get the status of an action.  The following should return the same status as the above `hasFinishedRequest`.
```
console.log( wp.data.select('experimental/wc/admin/products/categories').hasFinishedRequest('updateProductCategory',[{id:37}, {name: 'New name'}]) )
```
5. Generally smoke test the new product experience to make sure no regressions have occurred.
 


### Other information:




### FOR PR REVIEWER ONLY:
